### ISSUE TITLE
spack uninstall : cannot query by architecture or dependencies, just by compilers

### ISSUE THREAD

**Issue by alalazo** [LLNL/spack#1012]
Reproducible by : 

```
$ spack find
==> 33 installed packages.
-- linux-x86_64 / gcc@4.8 ---------------------------------------
binutils@2.26+gold~krellpatch~libiberty  gmp@6.1.0        libtool@2.4.6                                                                                                 m4@1.4.17+sigsegv  python@2.7.11  
bzip2@1.0.6                              isl@0.14         llvm@3.8.0+all_targets+clang+compiler-rt~debug+gold+internal_unwind+libcxx~link_dylib+lldb+polly~shared_libs  mpc@1.0.3          sqlite@3.8.5   
cmake@3.5.2~doc+ncurses+openssl~qt       libedit@3.1      lmod@6.0.1                                                                                                    mpfr@3.1.4         swig@3.0.8     
gcc@6.1.0+binutils+gold                  libsigsegv@2.10  lua@5.3.2                                                                                                     pcre@8.38          zlib@1.2.8     

-- linux-x86_64 / gcc@6.1.0 -------------------------------------
hdf5@1.10.0~cxx~debug+fortran+mpi+shared+szip~threadsafe  libsigsegv@2.10                     openmpi@1.10.2~mxm~pmi~psm~psm2~slurm~sqlite3~thread_multiple~tm~verbs+vt  zlib@1.2.8
hdf5@1.10.0+cxx~debug+fortran~mpi+shared+szip~threadsafe  libtool@2.4.6                       openmpi@1.10.2~mxm~pmi~psm~psm2~slurm+sqlite3+thread_multiple~tm~verbs+vt  
hwloc@1.11.3                                              m4@1.4.17+sigsegv                   sqlite@3.8.5                                                               
libpciaccess@0.13.4                                       openblas@0.2.18+fpic~openmp+shared  szip@2.1                                                                   

$ spack uninstall arch=linux-x86_64
==> Error: Unexpected end of input

$ spack uninstall =linux-x86_64
==> Error: Unexpected token
```

A similar issue will be triggered if querying for a dependency, e.g. : 

```
$ spack uninstall ^szip
==> Error: Unexpected end of input
```

Note that queries on compilers instead works : 

```
$ spack uninstall -a %gcc@6.1.0
==> The following packages will be uninstalled : 

-- linux-x86_64 / gcc@6.1.0 -------------------------------------
vhxsri5 hdf5@1.10.0%gcc~cxx~debug+fortran+mpi+shared+szip~threadsafe

q4cd33d hdf5@1.10.0%gcc+cxx~debug+fortran~mpi+shared+szip~threadsafe

whp7ocf hwloc@1.11.3%gcc

wwwjaz6 libpciaccess@0.13.4%gcc

tn76ik6 libsigsegv@2.10%gcc

rounbef libtool@2.4.6%gcc

m57favd m4@1.4.17%gcc+sigsegv

tvffito openblas@0.2.18%gcc+fpic~openmp+shared

fa3jbsy openmpi@1.10.2%gcc~mxm~pmi~psm~psm2~slurm~sqlite3~thread_multiple~tm~verbs+vt

7kcx5zj openmpi@1.10.2%gcc~mxm~pmi~psm~psm2~slurm+sqlite3+thread_multiple~tm~verbs+vt

4fjyj2l sqlite@3.8.5%gcc

prj7bjf szip@2.1%gcc

xfi4wik zlib@1.2.8%gcc


==> Do you want to proceed ? [y/n]
n
Operation aborted
```

**Comment by tgamblin** [comment]
@becker33

**Comment by adamjstewart** [comment]
@becker33 Is this fixed now?

**Comment by becker33** [comment]
It was once, but it seems to have been broken again. I added a fix to #2889 where I was already working on the parser.

**Comment by becker33** [comment]
Closing as fixed by #2889.
